### PR TITLE
Add optimistic locking

### LINKED ISSUE #3568: [SPIKE] Optimistic locking
Cocina objects are immutable (unlike Fedora objects), which leads to the following bug:
1. Service A loads cocina object 1.
2. Service A invokes Service B and passes cocina object 1.
3. Service B changes cocina object 1 by creating cocina object 1a with change X.
4. Service B persists cocina object 1a.
5. Service B completes and Service A resumes.
6. Service A change cocina object 1 by creating cocina object 1b with change Y (that is to an unrelated part of cocina object to change X).
7. Service A persists cocina object 1b <--- This overwrites change X.

One possible fix to this is optimistic locking:
1. Service A loads cocina object 1 at v1.
2. Service A invokes Service B and passes cocina object 1.
3. Service B changes cocina object 1 by creating cocina object 1a with change X.
4. Service B persists cocina object 1a. CocinaObjectStore checks that cocina object 1a was loaded from current version (v1) and when persisting changes to v2.
5. Service B completes and Service A resumes.
6. Service A change cocina object 1 by creating cocina object 1b with change Y (that is to an unrelated part of cocina object to change X).
7. Service A persists cocina object 1b. CocinaObjectStore checks that cocina object 1b was loaded from current version (v2). It wasn't, so raises an error.

Note that versions for the purpose of optimistic locking shouldn't be conflated with any of the other notions of versions. While a locking version can be an integer, it doesn't have to be.

### PR DESCRIPTION
**Pull request by jcoyne**
## Why was this change made? 🤔
Fixes #3568

## How was this change tested? 🤨

⚡ ⚠ If this change has cross service impact, including data writes to shared file system, ***run [integration tests](https://github.com/sul-dlss/infrastructure-integration-test)*** and/or test in [stage|qa] environment, in addition to specs. ⚡

### CODE REVIEW

**Comment by edsu** [comment]
Wow, the Rails support for ETag handling looks nice. Will rolling this out require any changes in clients like [dor-services](https://github.com/sul-dlss/dor-services)?

**Comment by jcoyne** [comment]
@edsu yes in dor-services-client:  https://github.com/sul-dlss/dor-services-client/pull/253

File: app/controllers/objects_controller.rb [hunk 1]
```diff
@@ -58,9 +58,17 @@ def create
   end
 
   def update
+    if request.headers['If-Match'] && (computed_etag(CocinaObjectStore.find(params[:id])) != request.headers['If-Match'])
+      return json_api_error(status: :precondition_failed,
+                            title: 'ETag mismatch',
+                            message: 'You are attempting to update a stale copy of the record. Refetch the record and attempt your change again.')
```
Comment (mjgiarlo) [inline app/controllers/objects_controller.rb:64]: Should this PR also update the OpenAPI Spec to reflect that some responses can be 412s?

**Review by mjgiarlo — APPROVED** [review]

File: app/controllers/objects_controller.rb [hunk 2]
```diff
@@ -73,8 +81,8 @@ def update
   def show
     (@cocina_object, created_at, modified_at) = CocinaObjectStore.find_with_timestamps(params[:id])
     headers['X-Created-At'] = created_at.httpdate
-    headers['Last-Modified'] = modified_at.httpdate
-    render json: @cocina_object
+    # Casting the Cocina object to a hash makes the etag stable.
+    render json: @cocina_object if stale?(etag: @cocina_object.to_h, last_modified: modified_at)
```
Comment (jmartin-sul) [inline app/controllers/objects_controller.rb:85]: so the call to `stale?` will automatically respond with a `304` if the caller's copy is not in fact stale?  that's my read of [the docs](https://apidock.com/rails/ActionController/ConditionalGet/stale%3F), but i'm new to `stale?` and `fresh_when`, and it wasn't totally clear to me from reading the `stale?` page.

**Review by jmartin-sul — APPROVED** [review]
a couple of questions (maybe suggestions if they're not off-base).  but this definitely seems like an improvement even if the questions/suggestions are worth taking up, so 👍

**Review by jmartin-sul — COMMENTED** [review]
also, what about a test for the `304` behavior for `ObjectsController#show`?

File: app/controllers/objects_controller.rb [hunk 3]
```diff
@@ -58,9 +58,17 @@ def create
   end
 
   def update
+    if request.headers['If-Match'] && (computed_etag(CocinaObjectStore.find(params[:id])) != request.headers['If-Match'])
+      return json_api_error(status: :precondition_failed,
```
[the remark this answers is not in the document]
Comment (jcoyne) [reply app/controllers/objects_controller.rb:62]: I don't think a transaction would prevent the case you described, however we could lock the row, but only once we switch to pg

File: app/controllers/objects_controller.rb [hunk 4]
```diff
@@ -58,9 +58,17 @@ def create
...
+                            title: 'ETag mismatch',
+                            message: 'You are attempting to update a stale copy of the record. Refetch the record and attempt your change again.')
+    end
+
     cocina_object = Cocina::Models.build(params.except(:action, :controller, :id).to_unsafe_h)
     updated_cocina_object = CocinaObjectStore.save(cocina_object)
 
+    # Casting the Cocina object to a hash makes the etag stable.
+    fresh_when(etag: cocina_object.to_h, last_modified: DateTime.now)
```
[the remark this answers is not in the document]
Comment (mjgiarlo) [reply app/controllers/objects_controller.rb:71]: If `updated_cocina_object` is a Cocina model instance, it may not respond to `:updated_at`.

**Comment by mjgiarlo** [comment]
> also, what about a test for the `304` behavior for `ObjectsController#show`?

This might be good to add to the OpenAPI spec as well.

File: app/controllers/objects_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (justinlittman) [reply app/controllers/objects_controller.rb:71]: We could return it from `save`.

[the remark this answers is not in the document]
Comment (mjgiarlo) [reply app/controllers/objects_controller.rb:71]: @justinlittman 💬 
> We could return it from `save`.

Sourced from Fedora? Yeah, we could totally do that.

[the remark this answers is not in the document]
Comment (jcoyne) [reply app/controllers/objects_controller.rb:71]: In an update situation, last_modified is going to be "right now", so why bother?  This is only used to index in dor_indexing_app.

File: app/controllers/objects_controller.rb (hunk 2 above)
Comment (jcoyne) [reply app/controllers/objects_controller.rb:85]: That's what the docs say.

**Comment by jcoyne** [comment]
> also, what about a test for the 304 behavior for ObjectsController#show?

I'm not inclined to do this because this is a built in behavior of Rails and we aren't depending on it.  The thing we depend on is that an ETag and a Last-Modified headers are created.  Both of these are tested.

File: openapi.yml [hunk 5]
```diff
@@ -1220,6 +1226,8 @@ paths:
             X-Created-At:
               schema:
                 type: string
+        '304':
+          description: No change. The client provided an ETag that matches the current state of the item.
```
Comment (mjgiarlo) [inline openapi.yml:1230]: 🍻 @jcoyne

**Comment by ndushay** [comment]
Should this PR be closed and branch deleted?

**Comment by ndushay** [comment]
> @ndushay why?

is this superseded by JLitt's work?

**Comment by jcoyne** [comment]
Uh, yep, looks like it.

**Pull request closed without merging**